### PR TITLE
BIP 134: Flexible Transactions

### PR DESCRIPTION
**Pull request by zander**
Introducing Flextrans. Requesting a BIP number assignment. Thank you!

### CODE REVIEW

**Comment by petertodd** [comment]
Wait, did you notice the license on this? That's not a free software compatible license due to the clause "Distribution of the work or derivative of the work in any standard (paper) book form is prohibited unless prior permission is obtained from the copyright holder."

In general, we probably should require authors to use a specific license to keep things simpler.

**Comment by MarcoFalke** [comment]
<sub>(slightly OT)</sub> BIP1 *requires* all BIPs to be either licensed according to the OPL or be placed in the public domain. Both are a bad choice in my opinion and even though BIP1 does not forbid dual-licensing the work, I'd very much prefer to explicitly have a wider choice of accepted licenses. I had some hope that BIP2 (#350) solves this, but unfortunately it was deferred.

**Comment by petertodd** [comment]
@MarcoFalke Yeah, I can't imagine Amir Taaki meant for the non-free, restrictive versions of the OPL to be used; with this BIP's licensing terms it wouldn't be possible for, say, Andreas to include it in the next book on the Bitcoin protocol that he writes. Nor would it be possible for a future version of Flexible Transactions to do the obvious thing and include/modify the text of this BIP without first obtaining the author's explicit permission.

For standards that need to be distributed widely, public domain (or equivalent) is much more appropriate.

**Comment by luke-jr** [comment]
I did notice the license, but it follows the rules in BIP 1 and doesn't legally impede the publication in the BIPs repository.

BIP 2 was held back by objections to the BIP Comments changes. I suppose I could try removing that and reopen it for discussion. Perhaps I ought to forbid the additional OPL restrictions in the process...

**Comment by luke-jr** [comment]
@zander Would you be okay with removing these restrictions from the BIP? Would you object to BIP 2 if it forbade such restrictions in the future?

File: bip-0134.mediawiki [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+In the compact message format we define tokens and in this specification we
+define how these tokens are named, where they can be placed and which are
+optional.  To refer to XML, this specification would be the schema of
+a transaction.
+
+CMF tokens are triplets of name, format (like PositiveInteger) and value.
+Names in this scope are defined much like an enumeration where the actual
+integer value (id, below) is equally important to the written name.
+If any token found that is not covered in the next table will make the
```
Comment (hoffmabc) [inline bip-0134.mediawiki]: If any token found that is not covered in the next table *it* will make the ...

Comment (hoffmabc) [inline bip-0134.mediawiki]: Missing the word 'it'.

File: bip-0134.mediawiki [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+|-
+|TxEnd         ||  0 ||BoolTrue ||  Required    ||A marker that is end of the transaction.
+|-
+|TxInPrevHash  ||  1 ||ByteArray||  Required    ||TxId we are spending
+|-
+|TxPrevIndex   ||  2 ||Integer  ||      0       ||Index in prev tx we are spending (applied to previous TxInPrevHash)
+|-
+|TxInScript    ||  3 ||ByteArray||  Required    ||The 'input' part of the script
+|-
+|TxOutValue    ||  4 ||Integer  ||  Required    ||Amount of satoshi to transfer
```
Comment (hoffmabc) [inline bip-0134.mediawiki]: Satoshi -> satoshis

File: bip-0134.mediawiki [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+has ended, allowing the next to be parsed.
+
+Notice that the token ScriptVersion is currently not allowed because we
+don't have any valid value to give it. But if we introduce a new script
+version it would be placed in the outputs segment.
+
+=== Script v2 ===
+
+The default value of ScriptVersion is number 2, as opposed to the version 1
+of script that the is in use today.  The version 2 is mostly identical
```
Comment (hoffmabc) [inline bip-0134.mediawiki]: Remove extraneous the.

File: bip-0134.mediawiki [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+The effect of leaving the signatures out of the calculation of the
+transaction-id implies that the signatures are also not used for the
+calculation of the merkle tree.  This means that changes in signatures
+would not be detectable. Except naturally by the fact that missing or
+broken signatures breaks full validation. But it is important to detect
+modifications to such signatures outside of validating all transactions.
+
+For this reason the merkle tree is extended to include (append) the hash of
+the v4 transactions. The markle tree will continue to have all the
+transactions' tx-ids but appended to that are the v4 hahes that include the
```
Comment (hoffmabc) [inline bip-0134.mediawiki]: Hahes -> hashes

File: bip-0134.mediawiki [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+signatures as well.  Specifically the hash is taken over a data-blob that
+is build up from:
```
Comment (hoffmabc) [inline bip-0134.mediawiki]: Built not build.

File: bip-0134.mediawiki [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+
+==Backwards compatibility ==
+
+Fully validating older clients are not compatible with this change.
+
+SPV (simple payment validation) wallets need to be updated to receive or
+create the new transaction type.
+
+This BIP introduces a new transaction format without changing or
+deprecating the existing one or any of its practices. Therefor it is
```
Comment (hoffmabc) [inline bip-0134.mediawiki]: Therefore misspelling.

**Comment by zander** [comment]
On Friday, 23 September 2016 15:46:09 CEST Luke-Jr wrote:
> @zander Would you be okay with removing these restrictions from the BIP?
> Would you object to BIP 2 if it forbade such restrictions in the future?

My suggestion would be that we replace OPL as an allowed license with one 
or two Creative Commons licenses. Following the suggestion from the OPL 
creators themselves.
According to Wikipedia;

> Open Publication License was created by the Open Content Project in 1999 
as public copyright license for documents.[1] The license was superseded in 
2003/2007 by the Creative commons licenses.[2]

I'd suggest saying that "Share alike" is required and "Attribution" is 
optional.

Executive summary; give the user the choice (next to public domain) between 
CCO and BY-SA
see;
https://en.wikipedia.org/wiki/
Creative_Commons_license#Seven_regularly_used_licenses

I would gladly change to one of those two licenses.

**Comment by luke-jr** [comment]
What I do right now is dual-license under the OPL *and* the 2-clause BSD license. You could also do OPL + some CC license immediately.

In the near future, I'm proposing BIP 2 replace BIP 1. It allows [both CC0 and CC-BY-SA as license options for BIPs](https://github.com/luke-jr/bips/blob/bip0002_squash/bip-0002.mediawiki#Recommended_licenses). I would appreciate if you could help review it. :)

**Comment by cpacia** [comment]
Regarding the spec.. it looks like a tag needs to be defined for `Sequence` (for RBF purposes). Probably only one per tx instead of per input?

Also, it's unclear how standard `nLockTime` would be handled. I assume `LockByTime` and `LockByBlock` are intended to be used for relative locktime (as they reference bip 68). They could also be treated as standard locktime if they are put in the `additional` part of the transaction. Otherwise a separate `nLockTime` tag should probably be defined.

And in regards to `LockByTime` and `LockByBlock` when used for relative locktime they should probably be added to the input sections no? For example:

Inputs = TxInPrevHash, TxInPrevIndex, LockByTime or LockByBlock. Similar to how `TxInPrevIndex` is set to zero if omitted, `LockByTime` and `LockByBlock` should also be treated as having a zero value if omitted. 

My 2¢

**Comment by zander** [comment]
> And in regards to LockByTime and LockByBlock when used for relative locktime they should probably be added to the input sections no?

The reason that the BIP68 data was added to the inputs was only because that was the only place in the transaction where there was space to use. Inputs are about *using* a former transactions output. The user of the output would not be the one that specifies how long that output has to be locked. You don't specify the rules of spending months after it confirms on the blockchain.

**Comment by cpacia** [comment]
> The user of the output would not be the one that specifies how long that output has to be locked. You don't specify the rules of spending months after it confirms on the blockchain.

When combined with CSV I believe it works that way.

**Pull request merged**